compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3633,3 +3634,43 @@ declare_lint! {
     Warn,
     "duplicated attribute"
 }
+
+declare_lint! {
+    /// The `not_cenum_cast` lint detects an `as` cast of a field-less
+    /// `enum` that is not C-like.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// enum E {
+    ///     A(),
+    ///     B{},
+    ///     C,
+    /// }
+    ///
+    /// fn main() {
+    ///     let i = E::A() as u32;
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Historically we permit casting of enums if none of the variants have
+    /// fields. The intended behaviour is that only C-like (all variants are
+    /// unit variants) enums can be casted this way.
+    ///
+    /// This is a [future-incompatible] lint to transition this to a hard error
+    /// in the future. See [issue #88621] for more details.
+    ///
+    /// [future-incompatible]: ../index.md#future-incompatible-lints
+    /// [issue #88621]: https://github.com/rust-lang/rust/issues/88621
+    pub NOT_CENUM_CAST,
```

Comment:
nit: `CENUM` looks a little off to me. Maybe `C_ENUM` would be better?

Reply:
There is an existing lint "CENUM_IMPL_DROP_CAST" and I just followed that.